商品情報編集機能

sasakikeisuke-ft opened the pull request:
# What
商品情報の編集機能

# why
出品後にも商品情報を修正できるようにするため

GyazoGIF
*[ログイン状態の出品者は、商品情報編集ページに遷移できる動画](https://gyazo.com/76533b4783d7145eb8a30002f5137ede)
*[正しく情報を記入すると、商品の情報を編集できる動画](https://gyazo.com/957207e98d04bda0276f182a81ae80ca)
*[入力に問題のある状態では商品の情報は編集できず、エラーメッセージが出力される動画](https://gyazo.com/da71b46fd8f1a22c661f88bcb19537d7)
*[何も編集せずに更新をしても画像無しの商品にならない動画](https://gyazo.com/0432538653b321d7ecb76f8750afe8f8)
*[ログイン状態のユーザーであっても、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画](https://gyazo.com/49cc7012dd4ffcc2a1536bf823ff5767)
*[ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画](https://gyazo.com/81249ebbf7a9c26df6438fa6d12fa734)
*[商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）](https://gyazo.com/ac15f234ea55ae0d4a9beb078c14178b)


本機能では初回のコードレビュー依頼です。
ご確認、ご指導をどうぞよろしくお願いします！

sasakikeisuke-ft committed:
editとuodateアクションへのルーティング設定

sasakikeisuke-ft committed:
itemsコントローラーにアクションを追加

itemsコントローラーにeditアクションとupdateアクションを追加しました。
またログインしているユーザーのみ該当アクションが行えるように制限しました。

sasakikeisuke-ft committed:
商品詳細ページに商品編集ページへのリンクを設定

sasakikeisuke-ft committed:
ログインしている出品者のみ編集を許可

ログインしているユーザーが出品者でない場合、トップページへ遷移するよう編集しました。

sasakikeisuke-ft committed:
商品編集ページの表示編集

商品編集ページに遷移されたら、各項目の元の情報がフォームに入力されている状態で表示されるよう編集しました。

sasakikeisuke-ft committed:
商品編集ページのエラーメッセージ編集

商品情報を変更した際に不適切な内容があるとエラーメッセージが出現するように編集しました。

sasakikeisuke-ft committed:
商品詳細ページへのリンク設定

ページ下部の戻るボタンから、元の商品詳細ページへ遷移するよう設定しました。

sasakikeisuke-ft committed:
実装予定の項目についてコメント

現在実装できない機能を今後実装するため、コメントを残しました。

sasakikeisuke-ft committed:
Rubocop実行

Rubocopを実行し、リファクタリングを行いました。
それに伴い、editアクションのunless文をreturn if へ変更しました。

ten0105 commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +22,22 @@ def show
   end
 
+  def edit
+    @item = Item.find(params[:id])
```
```ruby
@item = Item.find(params[:id])
```
@itemの記述が重複しているようです。
こちらはprivate以下にメソッド化して、before_actionで呼び出す形に修正しておきましょう。
複数回使用する変数をメソッド化することで可読性が上がり、改修作業が発生した際も手を加えやすくなります。

ten0105 commented on app/controllers/items_controller.rb:
```diff
@@ -22,6 +22,22 @@ def show
+    redirect_to root_path
+  end
+
+  def update
+    @item = Item.find(params[:id])
+    if @item.update(item_params)
+      redirect_to item_path(@item)
+    else
+      render :edit
+    end
+  end
```
editアクションと同様にupdateアクションにも以下の条件を追加しましょう。

```
ログアウト状態のユーザーが商品をアップデートしようとするとログインページへ遷移する。
出品者以外のユーザーが商品をアップデートしようとするとトップページヘ遷移する。
```

検証ツールを用いてlink_toのパスを変更することで、商品編集画面を経由せずにupdateアクションを動かすことが可能です。
そういった行為への対策として、updateアクションにもeditアクションと同様の条件分岐を加えることが必要となります。

ten0105 reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
お手数をお掛け致しますが、よろしくお願いいたします。

sasakikeisuke-ft committed:
修復1：重複するコードのメソッド化

editアクションとupdateアクションで重複していたコードをメソッド化し、before_acitonとして稼働するよう編集しました。

sasakikeisuke-ft committed:
修正2：出品者でないログインユーザーがupdateアクションを行うとトップページへ遷移する

指摘に基づき、出品者ではないログインユーザーがupdateアクションを行おうとした場合、トップページへ遷移するよう編集しました。
またログインユーザーと出品者が一致するか確認する挙動がeditアクションと同一出会ったため、privateメソッドとbefore_actionで設定しました。

sasakikeisuke-ft commented:
# What
商品情報の編集機能　第一次修正

# why
コードの可読性向上のため
意図しない挙動で出品者以外が商品情報を変更できないようにするため

GyazoGIF
動作には変更がないため割愛

ご指摘に基づき、修正しました！
ご指摘にある「ログアウト状態のユーザーが商品をアップデートしようとするとログインページへ遷移する」ですが、これはauthenticate_user!メソッドを使用しているため実装できていると考えます。
また「出品者以外のユーザーが商品をアップデートしようとするとトップページヘ遷移する」ですが、修正1で定義したfind_itemメソッドに合わせて、ログインユーザーと出品者が一致しているか検証するように設定しました。

自分なりの工夫になりますが、ご指摘いただければ幸いです。
よろしくお願いします！

s-aka-moto reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！

sasakikeisuke-ft commented:
確認ありがとうございます！
今後もご指導よろしくお願いします！
